File: strix-webui/backend/src/reports/chat-pdf-generator.ts
import PDFDocument from "pdfkit";
import type { ChatSession, ChatMessageRecord } from "@strix-webui/shared";

const COLORS = {
  bg: "#0A0A0A",
  text: "#FFFFFF",
  muted: "#A0A0A0",
  border: "#2A2A2A",
  accent: "#22C55E",
};

interface ParsedBlock {
  type: "text" | "tool";
  text?: string;
  tool?: { name: string; status: string; input: Record<string, unknown>; result?: string };
}

function formatDate(iso: string): string {
  return new Date(iso).toLocaleString("en-US", {
    year: "numeric", month: "short", day: "numeric",
    hour: "2-digit", minute: "2-digit",
  });
}

function ensureSpace(doc: PDFKit.PDFDocument, needed: number) {
  if (doc.y > doc.page.height - needed) {
    doc.addPage();
  }
}

/**
 * Normalize content that may have lost its newlines.
 * Inserts line breaks before markdown block markers that appear mid-line.
 */
function normalizeContent(text: string): string {
  // If text already has reasonable newlines, return as-is
  const lines = text.split("\n");
  if (lines.length > 3) return text;

  // Content might be a single long line — try to restore line breaks
  return text
    .replace(/([.!?:})\]]) (#{1,3} )/g, "$1\n\n$2")  // before headings
    .replace(/([.!?]) (- )/g, "$1\n$2")                 // before list items
    .replace(/(```)/g, "\n$1\n")                         // around code fences
    .replace(/\n{3,}/g, "\n\n");                         // collapse excess
}

/** Strip inline markdown and return plain text for safe PDFKit rendering */
function stripInlineMarkdown(text: string): string {
  return text
    .replace(/\*\*([^*]+)\*\*/g, "$1")
    .replace(/`([^`]+)`/g, "$1")
    .replace(/\*([^*]+)\*/g, "$1")
    .replace(/__([^_]+)__/g, "$1");
}

/** Render text content line-by-line (no `continued` to avoid PDFKit page-break crashes) */
function renderContent(doc: PDFKit.PDFDocument, text: string, pageWidth: number) {
  const normalized = normalizeContent(text);
  const lines = normalized.split("\n");
  let inCodeBlock = false;

  for (const line of lines) {
    ensureSpace(doc, 25);

    // Code block toggle
    if (line.trim().startsWith("```")) {
      inCodeBlock = !inCodeBlock;
      doc.moveDown(0.2);
      continue;
    }

    if (inCodeBlock) {
      doc.font("Courier").fontSize(8).fillColor(COLORS.accent)
        .text(line, { width: pageWidth });
      continue;
    }

    // Headings
    if (line.startsWith("#### ")) {
      doc.font("Helvetica-Bold").fontSize(10).fillColor(COLORS.text)
        .text(stripInlineMarkdown(line.slice(5)), { width: pageWidth });
      doc.moveDown(0.15);
      continue;
    }
    if (line.startsWith("### ")) {
      doc.font("Helvetica-Bold").fontSize(11).fillColor(COLORS.text)
        .text(stripInlineMarkdown(line.slice(4)), { width: pageWidth });
      doc.moveDown(0.2);
      continue;
    }
    if (line.startsWith("## ")) {
      doc.font("Helvetica-Bold").fontSize(12).fillColor(COLORS.text)
        .text(stripInlineMarkdown(line.slice(3)), { width: pageWidth });
      doc.moveDown(0.2);
      continue;
    }
    if (line.startsWith("# ")) {
      doc.font("Helvetica-Bold").fontSize(14).fillColor(COLORS.text)
        .text(stripInlineMarkdown(line.slice(2)), { width: pageWidth });
      doc.moveDown(0.3);
      continue;
    }

    // Horizontal rule
    if (/^-{3,}$/.test(line.trim()) || /^\*{3,}$/.test(line.trim())) {
      doc.moveTo(50, doc.y).lineTo(doc.page.width - 50, doc.y).stroke(COLORS.border);
      doc.moveDown(0.3);
      continue;
    }

    // Empty line
    if (!line.trim()) {
      doc.moveDown(0.25);
      continue;
    }

    // List items (- or * or numbered)
    if (/^\s*[-*]\s/.test(line)) {
      const indent = line.match(/^(\s*)/)?.[1]?.length || 0;
      const content = line.replace(/^\s*[-*]\s+/, "");
      doc.font("Helvetica").fontSize(10).fillColor(COLORS.text)
        .text(`${"  ".repeat(Math.floor(indent / 2))}  \u2022  ${stripInlineMarkdown(content)}`, { width: pageWidth });
      continue;
    }
    if (/^\s*\d+[.)]\s/.test(line)) {
      const content = line.replace(/^\s*(\d+[.)]\s+)/, "$1");
      doc.font("Helvetica").fontSize(10).fillColor(COLORS.text)
        .text(`  ${stripInlineMarkdown(content)}`, { width: pageWidth });
      continue;
    }

    // Blockquote
    if (line.startsWith("> ")) {
      doc.font("Helvetica-Oblique").fontSize(10).fillColor(COLORS.muted)
        .text(stripInlineMarkdown(line.slice(2)), { width: pageWidth - 20, indent: 20 });
      continue;
    }

    // Regular text — strip inline markdown, render as plain paragraph
    doc.font("Helvetica").fontSize(10).fillColor(COLORS.text)
      .text(stripInlineMarkdown(line), { width: pageWidth });
  }
}

export async function generateChatPDF(
  session: ChatSession,
  messages: ChatMessageRecord[],
): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const doc = new PDFDocument({
      size: "A4",
      margins: { top: 50, bottom: 60, left: 50, right: 50 },
      info: {
        Title: `Chat Export - ${session.title}`,
        Author: "Strix Security",
        Subject: "AI Chat Session Export",
      },
    });

    const chunks: Buffer[] = [];
    doc.on("data", (chunk: Buffer) => chunks.push(chunk));
    doc.on("end", () => resolve(Buffer.concat(chunks)));
    doc.on("error", reject);

    const pageWidth = doc.page.width - 100;

    // ==================== Cover Page ====================
    doc.moveDown(5);
    doc.fontSize(28).fillColor(COLORS.text).text("Chat Export", { align: "center" });
    doc.moveDown(0.5);
    doc.fontSize(16).fillColor(COLORS.accent).text(session.title, { align: "center", width: pageWidth });
    doc.moveDown(2);

    doc.fontSize(10).fillColor(COLORS.muted);
    doc.text(`Session created: ${formatDate(session.createdAt)}`, { align: "center" });
    doc.text(`Messages: ${messages.length}`, { align: "center" });
    if (messages.length > 0) {
      const first = formatDate(messages[0].createdAt);
      const last = formatDate(messages[messages.length - 1].createdAt);
      doc.text(`Time range: ${first} — ${last}`, { align: "center" });
    }
    doc.moveDown(6);
    doc.fontSize(9).fillColor(COLORS.muted).text("Exported from Strix", { align: "center" });

    // ==================== Messages ====================
    doc.addPage();

    let pageNum = 2;
    const addFooter = () => {
      doc.fontSize(8).fillColor(COLORS.muted)
        .text(`Exported from Strix — Page ${pageNum}`, 50, doc.page.height - 40, {
          align: "center", width: pageWidth,
        });
      pageNum++;
    };

    doc.on("pageAdded", addFooter);

    for (const msg of messages) {
      ensureSpace(doc, 50);

      const roleLabel = msg.role === "user" ? "User" : "Assistant";
      const roleColor = msg.role === "user" ? "#60A5FA" : COLORS.accent;
      const hasBlocks = !!msg.blocks;

      // Role + timestamp header
      doc.font("Helvetica-Bold").fontSize(11).fillColor(roleColor)
        .text(`${roleLabel}  `, { continued: true });
      doc.font("Helvetica").fontSize(9).fillColor(COLORS.muted)
        .text(formatDate(msg.createdAt));

      doc.moveDown(0.3);

      // For messages with blocks (execute mode), skip content to avoid duplication
      if (hasBlocks) {
        try {
          const blocks: ParsedBlock[] = JSON.parse(msg.blocks!);
          for (const block of blocks) {
            if (block.type === "text" && block.text) {
              renderContent(doc, block.text, pageWidth);
            } else if (block.type === "tool" && block.tool) {
              ensureSpace(doc, 35);
              const statusIcon = block.tool.status === "done" ? "OK"
                : block.tool.status === "error" ? "ERR" : "...";

              doc.font("Courier-Bold").fontSize(9).fillColor(COLORS.muted)
                .text(`Tool: ${block.tool.name} [${statusIcon}]`);

              const inputStr = JSON.stringify(block.tool.input);
              const truncInput = inputStr.length > 200 ? inputStr.slice(0, 200) + "..." : inputStr;
              doc.font("Courier").fontSize(7.5).fillColor(COLORS.muted)
                .text(`  Input: ${truncInput}`, { width: pageWidth });

              if (block.tool.result) {
                const truncResult = block.tool.result.length > 300
                  ? block.tool.result.slice(0, 300) + "..." : block.tool.result;
                doc.font("Courier").fontSize(7.5).fillColor(COLORS.muted)
                  .text(`  Result: ${truncResult}`, { width: pageWidth });
              }
              doc.moveDown(0.2);
            }
          }
        } catch { /* invalid blocks JSON */ }
      } else if (msg.content) {
        renderContent(doc, msg.content, pageWidth);
      }

      // Separator
      doc.moveDown(0.4);
      doc.moveTo(50, doc.y).lineTo(doc.page.width - 50, doc.y).stroke(COLORS.border);
      doc.moveDown(0.4);
    }

    if (messages.length === 0) {
      doc.fontSize(11).fillColor(COLORS.muted).text("No messages in this session.");
    }

    addFooter();
    doc.end();
  });
}

File: src-tauri/src/reports/chat_pdf_generator.rs
//! Chat session PDF export — port of `reports/chat-pdf-generator.ts`.
//!
//! Generates a PDF containing: cover page with session metadata,
//! then all messages rendered with basic markdown support
//! (headings, code blocks, lists, blockquotes, horizontal rules).

use crate::models::{ChatMessageRecord, ChatSession};
use super::pdf_writer::{format_date, normalize_content, strip_markdown, truncate_str, PdfWriter};

const COLOR_TEXT: &str = "#000000";
const COLOR_MUTED: &str = "#A0A0A0";
const COLOR_ACCENT: &str = "#22C55E";
const COLOR_USER: &str = "#60A5FA";


const TOOL_INPUT_TRUNCATE: usize = 200;
const TOOL_RESULT_TRUNCATE: usize = 300;

/// Parsed block from the `blocks` JSON field on assistant messages.
#[derive(serde::Deserialize)]
#[serde(tag = "type")]
enum ParsedBlock {
    #[serde(rename = "text")]
    Text { text: Option<String> },
    #[serde(rename = "tool")]
    Tool { tool: Option<ToolBlock> },
}

#[derive(serde::Deserialize)]
struct ToolBlock {
    name: String,
    status: Option<String>,
    input: Option<serde_json::Value>,
    result: Option<String>,
}

/// Generate a PDF export for a chat session.
pub fn generate_chat_pdf(session: &ChatSession, messages: &[ChatMessageRecord]) -> Vec<u8> {
    let mut pdf = PdfWriter::new(&format!("Chat Export - {}", session.title));

    // ════════════════════ Cover Page ════════════════════

    pdf.move_down(20.0);
    pdf.font_bold(28.0);
    pdf.color_hex(COLOR_TEXT);
    pdf.text_centered("Chat Export");
    pdf.move_down(3.0);

    pdf.font_regular(16.0);
    pdf.color_hex(COLOR_ACCENT);
    pdf.text_centered(&session.title);
    pdf.move_down(10.0);

    pdf.font_regular(10.0);
    pdf.color_hex(COLOR_MUTED);
    pdf.text_centered(&format!("Session created: {}", format_date(&session.created_at)));
    pdf.text_centered(&format!("Messages: {}", messages.len()));

    if !messages.is_empty() {
        let first = format_date(&messages[0].created_at);
        let last = format_date(&messages[messages.len() - 1].created_at);
        pdf.text_centered(&format!("Time range: {} - {}", first, last));
    }

    pdf.move_down(25.0);
    pdf.font_regular(9.0);
    pdf.color_hex(COLOR_MUTED);
    pdf.text_centered("Exported from Strix");

    // ════════════════════ Messages ════════════════════

    pdf.new_page();

    for msg in messages {
        pdf.ensure_space(20.0);

        // Role + timestamp header
        let role_label = if msg.role == "user" { "User" } else { "Assistant" };
        let role_color = if msg.role == "user" {
            COLOR_USER
        } else {
            COLOR_ACCENT
        };

        pdf.font_bold(11.0);
        pdf.color_hex(role_color);
        pdf.text(&format!("{}   {}", role_label, format_date(&msg.created_at)));
        pdf.move_down(1.5);

        // Render content
        let has_blocks = msg.blocks.is_some();

        if has_blocks {
            if let Some(ref blocks_json) = msg.blocks {
                if let Ok(blocks) = serde_json::from_str::<Vec<ParsedBlock>>(blocks_json) {
                    for block in &blocks {
                        match block {
                            ParsedBlock::Text { text: Some(text) } => {
                                render_markdown_content(&mut pdf, text);
                            }
                            ParsedBlock::Tool { tool: Some(tool) } => {
                                render_tool_block(&mut pdf, tool);
                            }
                            _ => {}
                        }
                    }
                }
            }
        } else if !msg.content.is_empty() {
            render_markdown_content(&mut pdf, &msg.content);
        }

        // Separator
        pdf.move_down(2.0);
        pdf.horizontal_line();
        pdf.move_down(2.0);
    }

    if messages.is_empty() {
        pdf.font_regular(11.0);
        pdf.color_hex(COLOR_MUTED);
        pdf.text("No messages in this session.");
    }

    pdf.finish()
}

/// Render markdown-ish text content line by line.
fn render_markdown_content(pdf: &mut PdfWriter, text: &str) {
    let normalized = normalize_content(text);
    let lines: Vec<&str> = normalized.split('\n').collect();
    let mut in_code_block = false;

    for line in lines {
        pdf.ensure_space(5.0);
        let trimmed = line.trim();

        // Code block toggle
        if trimmed.starts_with("```") {
            in_code_block = !in_code_block;
            pdf.move_down(1.0);
            continue;
        }

        if in_code_block {
            pdf.font_mono(8.0);
            pdf.color_hex(COLOR_ACCENT);
            pdf.text(line);
            continue;
        }

        // Headings
        if line.starts_with("#### ") {
            pdf.font_bold(10.0);
            pdf.color_hex(COLOR_TEXT);
            pdf.text(&strip_markdown(&line[5..]));
            pdf.move_down(0.5);
            continue;
        }
        if line.starts_with("### ") {
            pdf.font_bold(11.0);
            pdf.color_hex(COLOR_TEXT);
            pdf.text(&strip_markdown(&line[4..]));
            pdf.move_down(1.0);
            continue;
        }
        if line.starts_with("## ") {
            pdf.font_bold(12.0);
            pdf.color_hex(COLOR_TEXT);
            pdf.text(&strip_markdown(&line[3..]));
            pdf.move_down(1.0);
            continue;
        }
        if line.starts_with("# ") {
            pdf.font_bold(14.0);
            pdf.color_hex(COLOR_TEXT);
            pdf.text(&strip_markdown(&line[2..]));
            pdf.move_down(1.5);
            continue;
        }

        // Horizontal rule
        if is_horizontal_rule(trimmed) {
            pdf.horizontal_line();
            pdf.move_down(1.5);
            continue;
        }

        // Empty line
        if trimmed.is_empty() {
            pdf.move_down(1.5);
            continue;
        }

        // List items (unordered)
        if trimmed.starts_with("- ") || trimmed.starts_with("* ") {
            let content = &trimmed[2..];
            pdf.font_regular(10.0);
            pdf.color_hex(COLOR_TEXT);
            pdf.text(&format!("  \u{2022}  {}", strip_markdown(content)));
            continue;
        }

        // List items (ordered)
        if is_ordered_list_item(trimmed) {
            pdf.font_regular(10.0);
            pdf.color_hex(COLOR_TEXT);
            pdf.text(&format!("  {}", strip_markdown(trimmed)));
            continue;
        }

        // Blockquote
        if line.starts_with("> ") {
            pdf.font_italic(10.0);
            pdf.color_hex(COLOR_MUTED);
            pdf.text(&format!("    {}", strip_markdown(&line[2..])));
            continue;
        }

        // Regular text
        pdf.font_regular(10.0);
        pdf.color_hex(COLOR_TEXT);
        pdf.text(&strip_markdown(line));
    }
}

/// Render a tool use block (name, status, truncated input/result).
fn render_tool_block(pdf: &mut PdfWriter, tool: &ToolBlock) {
    pdf.ensure_space(12.0);

    let status_icon = match tool.status.as_deref() {
        Some("done") => "OK",
        Some("error") => "ERR",
        _ => "...",
    };

    pdf.font_bold(9.0);
    pdf.color_hex(COLOR_MUTED);
    pdf.text(&format!("Tool: {} [{}]", tool.name, status_icon));

    if let Some(ref input) = tool.input {
        let input_str = serde_json::to_string(input).unwrap_or_default();
        let truncated = truncate_str(&input_str, TOOL_INPUT_TRUNCATE);
        pdf.font_mono(7.5);
        pdf.color_hex(COLOR_MUTED);
        pdf.text(&format!("  Input: {}", truncated));
    }

    if let Some(ref result) = tool.result {
        let truncated = truncate_str(result, TOOL_RESULT_TRUNCATE);
        pdf.font_mono(7.5);
        pdf.color_hex(COLOR_MUTED);
        pdf.text(&format!("  Result: {}", truncated));
    }

    pdf.move_down(1.0);
}

fn is_horizontal_rule(s: &str) -> bool {
    (s.len() >= 3 && s.chars().all(|c| c == '-'))
        || (s.len() >= 3 && s.chars().all(|c| c == '*'))
}

fn is_ordered_list_item(s: &str) -> bool {
    s.chars()
        .take_while(|c| c.is_ascii_digit())
        .count()
        > 0
        && (s.contains(". ") || s.contains(") "))
}
